### PR TITLE
add reverse_transcribe to RNA class

### LINKED ISSUE #1052: Add `reverse_transcribe` to `RNA`
It is missing.

### PR DESCRIPTION
**Pull request by RNAer**
Fix #1052

### CODE REVIEW

**Comment by RNAer** [comment]
@ebolyen , I just went ahead to solve the cyclic imports by putting them at the end as I can't make the `import` work. feel free to modify it to whatever your import style is.

**Comment by ebolyen** [comment]
Thanks @RNAer, I'll pull locally and see if I can work out the imports

File: skbio/sequence/_rna.py [hunk 1]
```diff
@@ -136,6 +136,72 @@ def _motifs(self):
...
+        metadata = None
+        if self.has_metadata():
+            metadata = self.metadata
+
+        positional_metadata = None
+        if self.has_positional_metadata():
+            positional_metadata = self.positional_metadata
+
+        # turn off validation because `seq` is guaranteed to be valid
+        return DNA(seq, metadata=metadata,
```
Comment (ebolyen) [inline skbio/sequence/_rna.py]: return `skbio.DNA`

File: skbio/sequence/_rna.py [hunk 2]
```diff
@@ -348,3 +414,5 @@ def _repr_stats(self):
 
 # Leave this at the bottom
 _motifs.interpolate(RNA, "find_motifs")
+
+from ._dna import DNA  # noqa
```
Comment (ebolyen) [inline skbio/sequence/_rna.py]: remove

File: skbio/sequence/_dna.py [hunk 3]
```diff
@@ -408,3 +407,5 @@ def _repr_stats(self):
 
 # Leave this at the bottom
 _motifs.interpolate(DNA, "find_motifs")
+
+from ._rna import RNA  # noqa
```
Comment (ebolyen) [inline skbio/sequence/_dna.py]: remove

File: skbio/sequence/_dna.py [hunk 4]
```diff
@@ -7,10 +7,8 @@
 # ----------------------------------------------------------------------------
 
 from __future__ import absolute_import, division, print_function
-
```
Comment (ebolyen) [inline skbio/sequence/_dna.py]: leave newline, add: `import skbio`

**Comment by ebolyen** [comment]
In `DNA.transcribe` change the constructor to `skbio.RNA`

**Comment by ebolyen** [comment]
@RNAer My above comments should correct the cyclic import without need for `# noqa`

File: skbio/sequence/tests/test_rna.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+import unittest
+
+from skbio import DNA, RNA
+
+
+# tests specific to RNA go here. tests for functionality shared by DNA and RNA
+# go in test_nucleotide_sequences.py
+class TestRNA(unittest.TestCase):
+    def test_transcribe(self):
```
Comment (ebolyen) [inline skbio/sequence/tests/test_rna.py]: `test_reverse_transcribe`

File: skbio/sequence/tests/test_rna.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        self.assertEqual(RNA('A').reverse_transcribe(), DNA('A'))
+        self.assertEqual(RNA('.ACGW-').reverse_transcribe(), DNA('.ACGW-'))
+
+        # with changes
+        self.assertEqual(DNA('T'), RNA('U').reverse_transcribe())
+        self.assertEqual(DNA('TT'), RNA('UU').reverse_transcribe())
+        self.assertEqual(DNA('ATCTG'), RNA('AUCUG').reverse_transcribe())
+        self.assertEqual(DNA('TTTG'), RNA('UUUG').reverse_transcribe())
+
+    def test_transcribe_preserves_all_metadata(self):
```
Comment (ebolyen) [inline skbio/sequence/tests/test_rna.py]: add `reverse` in there

File: skbio/sequence/tests/test_rna.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        seq = RNA('AGUU', metadata={'foo': 'bar'},
+                  positional_metadata={'foo': range(4)})
+        exp = DNA('AGTT', metadata={'foo': 'bar'},
+                  positional_metadata={'foo': range(4)})
+        self.assertEqual(seq.reverse_transcribe(), exp)
+
+    def test_transcribe_does_not_modify_input(self):
```
Comment (ebolyen) [inline skbio/sequence/tests/test_rna.py]: add `reverse` in there

File: skbio/sequence/_rna.py [hunk 8]
```diff
@@ -136,6 +136,72 @@ def _motifs(self):
...
+
+        See Also
+        --------
+        DNA.transcribe
+        translate
+        translate_six_frames
+
+        Notes
+        -----
+        DNA sequence's metadata and positional metadata are included in the
```
Comment (ebolyen) [inline skbio/sequence/_rna.py]: This note wasn't updated. It is going the wrong direction.

File: skbio/sequence/_rna.py [hunk 9]
```diff
@@ -136,6 +136,72 @@ def _motifs(self):
         return _motifs
 
     @stable(as_of="0.4.0")
```
Comment (ebolyen) [inline skbio/sequence/_rna.py]: I agree that this is stable, but make it "0.4.0-dev". When we release we will figure out what we are doing with the versions.

File: skbio/sequence/_dna.py [hunk 10]
```diff
@@ -222,6 +220,7 @@ def translate(self, *args, **kwargs):
 
         See Also
         --------
+        RNA.reverse_transcribe
```
Comment (jairideout) [inline skbio/sequence/_dna.py]: Can you remove this line and `transcribe`? I think I accidentally added `transcribe` when writing this code. The see also block should just point to translation methods.

File: skbio/sequence/_rna.py [hunk 11]
```diff
@@ -136,6 +136,72 @@ def _motifs(self):
...
+
+        Returns
+        -------
+        DNA
+            Reverse transcribed sequence.
+
+        See Also
+        --------
+        DNA.transcribe
+        translate
```
Comment (jairideout) [inline skbio/sequence/_rna.py]: Remove this line and `translate_six_frames`, see note above for reasoning

File: skbio/sequence/_rna.py [hunk 12]
```diff
@@ -136,6 +136,72 @@ def _motifs(self):
...
+    def reverse_transcribe(self):
+        """Reverse transcribe RNA into DNA.
+
+        Uracil (U) is replaced with thymine (T) in the reverse transcribed
+        sequence.
+
+        Returns
+        -------
+        DNA
+            Reverse transcribed sequence.
```
Comment (jairideout) [inline skbio/sequence/_rna.py]: Can you add `(coding strand)` to the end of this description? We make note of this in `DNA.transcribe`.

File: skbio/sequence/_rna.py (hunk 8 above)
Comment (jairideout) [inline skbio/sequence/_rna.py]: This needs to be framed in terms of reverse transcription.

Comment (jairideout) [inline skbio/sequence/_rna.py]: Woops, hadn't refreshed my page in awhile... sorry for the duplicate comment.

**Comment by ebolyen** [comment]
@RNAer Looks good for merge after addressing these comments.

File: skbio/sequence/_dna.py (hunk 10 above)
Comment (RNAer) [inline skbio/sequence/_dna.py]: @jairideout , why remove this line? it is related to `RNA.reverse_transcribe`.

Comment (ebolyen) [inline skbio/sequence/_dna.py]: I would vote to keep, these are all related to the central dogma (roughly). Talked with @jairideout offline and he's fine with keeping it also.

**Comment by RNAer** [comment]
:+1: They should be all addressed and ready to merge.

File: skbio/sequence/_rna.py [hunk 13]
```diff
@@ -135,6 +135,73 @@ def degenerate_map(cls):
     def _motifs(self):
         return _motifs
 
+    @stable(as_of="0.4.0-dev")
+    def reverse_transcribe(self):
+        """Reverse transcribe RNA into DNA.
+
+        RNA sequence is assumed to be the coding strand. Uracil (U) is
```
Comment (ebolyen) [inline skbio/sequence/_rna.py]: This doesn't make a lot of sense, the returned `DNA` is the coding strand, not really the `RNA` which is assumed to be `mRNA` in this context.

File: skbio/sequence/_rna.py (hunk 12 above)
Comment (ebolyen) [inline skbio/sequence/_rna.py]: I think the parenthetical on the return description is really all we need. It's just to confirm the common sense assumption about the return.

**Comment by ebolyen** [comment]
@RNAer still a slight problem with the docstring of `reverse_transcribe`, but otherwise looks good.

Can you also add a changelog mention?

File: skbio/sequence/_rna.py (hunk 13 above)
Comment (RNAer) [inline skbio/sequence/_rna.py]: not neccessarily. There can also be cases of that I have a ncRNA transcript
and want to get its DNA strand. That was why I didn't have "coding strand"
in the docstring previously... but it seems there are no good terminology
for that. Per @jarideout 's suggestion,  I can see it is useful to
distinguish coding strand and template strand. How about this "It returns
the coding DNA strand of the RNA sequence, i.e. Uracil (U) is replaced with
thymine (T) in the reverse transcribed sequence."?

Comment (RNAer) [inline skbio/sequence/_rna.py]: or just replace "coding strand" with "non-template strand"?

Comment (ebolyen) [inline skbio/sequence/_rna.py]: I like the first suggestion.

**Comment by RNAer** [comment]
@ebolyen , ok now? It should pass the checks.

File: skbio/sequence/_rna.py (hunk 13 above)
Comment (RNAer) [inline skbio/sequence/_rna.py]: actually we probably use the this terminology (
https://en.wikipedia.org/wiki/Coding_strand):

Wherever a gene <https://en.wikipedia.org/wiki/Gene> exists on a DNA
molecule <https://en.wikipedia.org/wiki/Molecule>, one strand is the coding
strand (or *sense strand* or *non-template strand*), and the other is
the *noncoding
strand* (also called the antisense
<https://en.wikipedia.org/wiki/Sense_%28molecular_biology%29> strand,[1]
<https://en.wikipedia.org/wiki/Coding_strand#cite_note-Lewin08-1>
anticoding strand, template strand, or transcribed strand).

I prefer the use of "sense" and "antisense" to accommandate both
protein-coding and non-protein-coding genes...

Comment (ebolyen) [inline skbio/sequence/_rna.py]: :+1: :heavy_multiplication_x: :100: for `sense` and `antisense` that is much more precise.

**Comment by ebolyen** [comment]
@RNAer Looks good! Let's just create an issue for renaming things to `sense` and `antisense` there is more than one place where we use `coding` and `noncoding`

File: skbio/sequence/_rna.py (hunk 13 above)
Comment (RNAer) [inline skbio/sequence/_rna.py]: cool. I will update docstrings for both transcribe() and
reverser_transcribe()

**Pull request merged**